chromadb/ingest/__init__.py

```diff
@@ -67,7 +71,43 @@ def max_batch_size(self) -> int:
         pass
 
 
-ConsumerCallbackFn = Callable[[Sequence[LogRecord]], None]
+# The consumer callback receives the subscription object so that it can call .ack() on it.
+# Ideally this method would be on the Consumer interface instead—why is on the subscription object?
+# The log stream enforces this lifecycle:
+#   subscription created ->
+#   backfill -> (subscription callback using below definition) ->
+#   register subscription on stream
+#
+# So during backfill, if .ack() was instead on the Consumer interface, the provided
+# callback wouldn't have a way to call .ack() as the subscription isn't registered until after backfilling.
+ConsumerCallbackFn = Callable[[Sequence[LogRecord], "Subscription"], None]
```

Comment:
wondering if there's a better approach, the alternative is requiring the callback to return the seq ID:

```suggestion
ConsumerCallbackFn = Callable[[Sequence[LogRecord]], SeqID]
```

however, this means the callback can't report partial work (e.x. the first 5 records are processed successfully, then on the 6th record it throws an error)

Reply:
I've had an impl in some of the PRs where each segment was responsible for updating its MaxSeqId in sqlite `max_seq_id`. Which can be used in cleaning. This will require hnsw to get the sqlite and update its maxSeqId when data is written to disk.

Reply:
interesting, could you link me? I think the PR I saw was roughly using the callback approach described above

Reply:
@codetheweb  https://github.com/chroma-core/chroma/pull/1868/files, check the `_persist` function in `local_persistent_hnsw.py`.

Reply:
I do like that approach a little more for its simplicity, would also avoid the issue of having to load segments into memory before pruning can take place
@HammadB and I discussed something somewhat similar yesterday
I'll take a look and see if it's feasible